pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -158,26 +170,288 @@ func New(cfg Config,
+				continue
+			}
+
+			tables = append(tables, table)
+			seen[table] = struct{}{}
+		}
+	}
+
+	// process most recent tables first
+	sortTablesByRange(tables)
+
+	// apply passed in compaction limits
+	if c.cfg.SkipLatestNTables <= len(tables) {
+		tables = tables[c.cfg.SkipLatestNTables:]
+	}
+	if c.cfg.TablesToCompact > 0 && c.cfg.TablesToCompact < len(tables) {
+		tables = tables[:c.cfg.TablesToCompact]
+	}
+
+	// Reset discovered tenants metric since we will increase it in compactTable.
+	c.metrics.compactionRunDiscoveredTenants.Set(0)
+
+	parallelism := c.cfg.MaxCompactionParallelism
+	if parallelism == 0 {
+		parallelism = len(tables)
+	}
+
+	errs := multierror.New()
+	if err := concurrency.ForEachJob(ctx, len(tables), parallelism, func(ctx context.Context, i int) error {
+		tableName := tables[i]
+		level.Info(util_log.Logger).Log("msg", "compacting table", "table-name", tableName)
+		err := c.compactTable(ctx, tableName)
+		if err != nil {
+			errs.Add(err)
+			return nil
+		}
+		level.Info(util_log.Logger).Log("msg", "finished compacting table", "table-name", tableName)
+		return nil
+	}); err != nil {
+		errs.Add(err)
+	}
+
+	return errs.Err()
+}
+
+func (c *Compactor) compactTable(ctx context.Context, tableName string) error {
+	schemaCfg, ok := schemaPeriodForTable(c.schemaCfg, tableName)
+	if !ok {
+		level.Error(util_log.Logger).Log("msg", "skipping compaction since we can't find schema for table", "table", tableName)
+		return nil
+	}
+
+	sc, ok := c.storeClients[schemaCfg.From]
+	if !ok {
+		return fmt.Errorf("index store client not found for period starting at %s", schemaCfg.From.String())
+	}
+
+	_, tenants, err := sc.index.ListFiles(ctx, tableName, false)
+	if err != nil {
+		return fmt.Errorf("failed to list files for table %s: %w", tableName, err)
+	}
+
+	c.metrics.compactionRunDiscoveredTenants.Add(float64(len(tenants)))
+	level.Info(c.logger).Log("msg", "discovered tenants from bucket", "users", len(tenants))
+	return c.compactUsers(ctx, sc, tableName, tenants)
+}
+
+// See: https://github.com/grafana/mimir/blob/34852137c332d4050e53128481f4f6417daee91e/pkg/compactor/compactor.go#L566-L689
+func (c *Compactor) compactUsers(ctx context.Context, sc storeClient, tableName string, tenants []string) error {
+	// When starting multiple compactor replicas nearly at the same time, running in a cluster with
+	// a large number of tenants, we may end up in a situation where the 1st user is compacted by
+	// multiple replicas at the same time. Shuffling users helps reduce the likelihood this will happen.
+	rand.Shuffle(len(tenants), func(i, j int) {
+		tenants[i], tenants[j] = tenants[j], tenants[i]
+	})
+
+	// Keep track of tenants owned by this shard, so that we can delete the local files for all other users.
+	errs := multierror.New()
+	ownedTenants := make(map[string]struct{}, len(tenants))
+	for _, tenant := range tenants {
+		// Ensure the context has not been canceled (ie. compactor shutdown has been triggered).
```

Comment:
You have a mix of `tenant` and `tenantID` for the tenant field in the logs.
To avoid that, you can create a logger with the field upfront:

```go
logger := log.With(c.logger, "tenant", tenant)
```

Reply:
Awesome, thanks!